### DISCUSSION TITLE
Make Storage Adapters an extension type

### DISCUSSION THREAD

**Discussion by rijkvanzanten** [directus/directus#24891] [Feature Requests]
With #16825 merged and released, we can take a next big step and unlock storage adapters as an extension type! This should in turn allow the community to make their own drivers for third party providers that don't meet 80/20 for core.

The implementation in the API should be relatively straightforward. Right now, this util function throws an error when an invalid name is used:

https://github.com/directus/directus/blob/dc1a218c58110fc28ec58a43b817069ec72df173/api/src/storage/get-storage-driver.ts#L12-L20

If we replace that with an async check to see if that name exists as a node module and then async load that in like we are with the current ones, we unlock npm-installable storage adapters.

The biggest lift to really support this is documentation.

**Comment by rijkvanzanten** [comment]
For those following along!

The storage adapter setup is already built modular, and Directus' native supported drivers are actually already implemented as "extensions" 🙂 That being said, I'd really like the storage driver setup to be a bit _better_ (think matching folders, making sure Directus can read files that don't exist in directus_files yet, etc). That would be a massive breaking change in the storage adapter setup, which means I'm a little conflicted on whether or not to open it up to custom extensions now, or wait for those breaking changes to be out the door.

**Comment by hiendaovinh** [comment]
I would like to store my files in different table rather than directus_files but fully accessible and rendered as image from the CMS.
Any suggestion for now?
Thank you.

**Comment by customautosys** [comment]
Hi any update on this?

I would like to try my hand at writing a storage adapter that stores it in a table in the same database using a blob column. This is due to the inode limit on my shared hosting and also due to likely migration of another database that stores images in blob columns.

**Comment by rijkvanzanten** [comment]
No updates yet. It's currently a backlog item as I'm a tad hesitant to open the floodgates for extensions on this point before doing an upgrade to how files are handled. There's a couple of things I'd like to tweak in the file abstraction in general—like how it syncs with the disk, where it stores thumbnails, and how it handles batch deletes—before we make it an extension type.
